gobblin-utility/src/main/java/gobblin/broker/iface/GobblinBroker.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.broker.iface;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import com.typesafe.config.Config;
+
+
+/**
+ * A class that provides access to objects shared by multiple components within a process, as well as objects virtually
+ * shared among different processes (i.e. objects that synchronize with equivalent objects in other processes).
+ *
+ * Different parts of an application may require objects that are ideally shared. For example, if multiple objects are
+ * writing to the same file, it is convenient to have a single file handle. Instead of passing around all shared objects,
+ * using static objects, or doing dependency injection, a {@link GobblinBroker} provides a way to acquire such shared
+ * objects as needed, letting the broker manage the lifecycle of the objects.
+ *
+ * Objects are created using {@link SharedResourceFactory}s, and are specific to a particular {@link Scope} and
+ * {@link SharedResourceKey}. {@link Scope}s represent a DAG of relevant scopes in the application, for example
+ * GLOBAL -> JOB -> TASK. {@link SharedResourceKey} identify different objects created by the same factory, for example
+ * handles to different files could be keyed by the file path.
+ *
+ * {@link GobblinBroker} guarantees that multiple requests for objects with the same {@link SharedResourceFactory},
+ * {@link Scope} and {@link SharedResourceKey} return the same object, even if called for different {@link GobblinBroker}
+ * instances (as long as the instances were created in the way specified by the {@link GobblinBroker} implementation).
+ *
+ * @param <S> the {@link ScopeType} tree used by this {@link GobblinBroker}.
+ */
+public interface GobblinBroker<S extends Enum<S> & ScopeType<S>> extends Closeable {
+
+  /**
+   * Get a {@link GobblinBroker} at the specified scope, used to get shared objects at that scope.
+   * Note that requesting a {@link ScopeType} that is not an ancestor of the {@link #leafScope()} will throw a
+   * {@link NoSuchScopeException}.
+   *
+   * @param scopeType the {@link ScopeType} requested.
+   * @return a {@link GobblinBroker} at the specified scope
+   * @throws NoSuchScopeException
+   */
+  ScopedGobblinBroker<S> scopedBroker(S scopeType) throws NoSuchScopeException;
+
+  /**
+   * @return The lowest {@link Scope} for which this {@link GobblinBroker} can provide a {@link ScopedGobblinBroker}.
+   */
+  Scope<S> leafScope();
```

Comment:
Is `leafScope()` used for the callers to check if the requested `scopeType` is provided? Then why not simply provide a collection of all the `scopeType`s that are supported?